pkg/util/notification.go

```diff
@@ -38,4 +47,66 @@ const (
 
 	// AnnotationsPrefix is the notification annotations prefix
 	AnnotationsPrefix = "notifications." + configsync.GroupName
+
+	// SubscribeAnnotationPrefix is the annotation prefix of the notification subscription.
+	SubscribeAnnotationPrefix = AnnotationsPrefix + "/subscribe."
+
+	// MultiSubscriptionsField is the field name for subscriptions configured globally.
+	MultiSubscriptionsField = "subscriptions"
+
+	// NotifiedAnnotationPath is the annotation path that indicates it has been notified.
+	NotifiedAnnotationPath = ".metadata.annotations.notified." + AnnotationsPrefix
 )
+
+// NotificationEnabled returns whether the notification is enabled for the RSync object.
+func NotificationEnabled(ctx context.Context, client client.Client, rsNamespace string, annotations map[string]string, notificationConfig *v1beta1.NotificationConfig) (bool, error) {
+	hasSubscriptionAnnotation := false
+	// subscription via prefix
+	for key := range annotations {
+		if strings.HasPrefix(key, SubscribeAnnotationPrefix) {
+			hasSubscriptionAnnotation = true
+		}
+	}
+	if hasSubscriptionAnnotation && notificationConfig == nil {
+		return false, fmt.Errorf("subscription annotation was provided without a notificationConfig")
+	}
+	if notificationConfig == nil {
+		return false, nil
+	}
+	// ConfigMapRef must be defined
+	configMapName := v1beta1.GetConfigMapName(notificationConfig.ConfigMapRef)
```

Comment:
If there is no subscription annotation, should this be allowed?

Reply:
My thinking was that `notificationConfig` should be empty (`nil`) if the user is not enabling the notification feature. And if the user _is_ enabling the notification feature, the `configMapRef` _must_ be set.

Are you suggesting that we should return `false, nil` if `notificationConfig` is not `nil` but `configMapRef` is unset/empty?

Reply:
My thought is `if !hasSubscriptionNotification && notificationConfig != nil`, we can return `false, nil` no matter if `configMapRef` is set/empty or not.

Basically, if the subscription is not enabled, we don't need to validate the notificationConfig content.
One example is that a user originally enabled the subscription with configMapRef, then the user unsubscribed, and the ConfigMap was deleted from the cluster. We shouldn't return a validation error for such case, because the ConfigMap is not needed anymore.

Reply:
The only issue I see with that is the ConfigMap-based default subscriptions (set via the `subscriptions` field). A valid/enabled notification configuration could have the ConfigMap set with default subscriptions but no annotations

Reply:
If the user would like to use the ConfigMap-based subscriptions, he/she should configure the configMapRef. If it is not configured, we can just treat it like disabled.

Reply:
done